oximeter/instruments/src/kstat/link.rs

```diff
@@ -0,0 +1,435 @@
+                    .take(5)
+                    .map(char::from)
+                    .collect::<String>(),
+            );
+            Self::create(&name);
+            Self { name }
+        }
+
+        fn create(name: &str) {
+            let output = std::process::Command::new(Self::PFEXEC)
+                .env_clear()
+                .arg(Self::DLADM)
+                .arg("create-etherstub")
+                .arg("-t")
+                .arg(name)
+                .output()
+                .expect("failed to spawn dladm");
+            assert!(
+                output.status.success(),
+                "failed to create test etherstub:\n{}",
+                String::from_utf8_lossy(&output.stderr)
+            );
+        }
+    }
+
+    impl Drop for TestEtherstub {
+        fn drop(&mut self) {
+            let output = std::process::Command::new(Self::PFEXEC)
+                .env_clear()
+                .arg(Self::DLADM)
+                .arg("delete-etherstub")
+                .arg(&self.name)
+                .output()
+                .expect("failed to spawn dladm");
+            if !output.status.success() {
+                eprintln!(
+                    "Failed to delete etherstub '{}'.\n\
+                    Delete manually with `dladm delete-etherstub {}`:\n{}",
+                    &self.name,
+                    &self.name,
+                    String::from_utf8_lossy(&output.stderr),
+                );
+            }
+        }
+    }
+
+    #[test]
+    fn test_physical_datalink() {
+        let link = TestEtherstub::new();
+        let sn = String::from("BRM000001");
+        let dl = PhysicalDataLink {
+            rack_id: RACK_ID,
+            sled_id: SLED_ID,
+            serial: sn.clone(),
+            hostname: sn,
+            link_name: link.name.to_string(),
+        };
+        let ctl = Ctl::new().unwrap();
+        let ctl = ctl.update().unwrap();
+        let mut kstat = ctl
+            .filter(Some("link"), Some(0), Some(dl.link_name.as_str()))
+            .next()
+            .unwrap();
+        let data = ctl.read(&mut kstat).unwrap();
+        let samples = dl.to_samples(&[(Utc::now(), kstat, data)]).unwrap();
+        println!("{samples:#?}");
+    }
+
+    #[tokio::test]
+    async fn test_kstat_sampler() {
+        let mut sampler = KstatSampler::new(&test_logger()).unwrap();
+        let sn = String::from("BRM000001");
+        let link = TestEtherstub::new();
+        let dl = PhysicalDataLink {
+            rack_id: RACK_ID,
+            sled_id: SLED_ID,
+            serial: sn.clone(),
+            hostname: sn,
+            link_name: link.name.to_string(),
+        };
+        let details = CollectionDetails::never(Duration::from_secs(1));
+        let id = sampler.add_target(dl, details).await.unwrap();
+        let samples: Vec<_> = sampler.produce().unwrap().collect();
+        assert!(samples.is_empty());
+        tokio::time::sleep(Duration::from_secs(2)).await;
```

Comment:
I'm not a huge fan of sleeps in tests, but removing these seems a little challenging. I think the way I might try to do it here would be

- make `KstatSamplerWorker::run` take an `Option<$TOKIO_SYNC_PRIMITIVE>` (haven't decided on exactly what kind yet); if this is `Some` then the worker signals/pokes/etc. the widget when it actually decides to sample something
- add a `#[cfg(test)]` object to `KstatSampler` that holds the 'receiver' end of the sync widget
- use [`tokio::time::pause`](https://docs.rs/tokio/latest/tokio/time/fn.pause.html) and `advance` to drive the clock instead of sleeping
- have the test make sure that once the clock is driven, the sample worker eventually pokes the widget

The extra "I did something" signal is needed because `advance` only guarantees that time advances and not that any specific future has been polled or reached a subsequent await point.

It's some extra code to carry around, but it will make the tests run more quickly, and I think it also lets you get rid of the Boolean OR on line 359--you can control exactly how much time advances, and you know when the sampler has run, so you can now reason about exactly how many samples it should have picked up.

I think this would also be applicable to the other tests in this module that sleep.

WDYT?

Reply:
Yeah, that's definitely a more principled approach to things. I've used that with relative success in other code, so I'll give it a whirl here too. This was the quick-and-dirty version :)

Reply:
Ok this was all added in 430df98d3. It's a bit gross, TBH, since manually stepping time requires that we use `tokio::time::Instant` in place of a `DateTime<Utc>`. That's because things like `tokio::time::advance` just move Tokio's internal clock, while UTC is obviously unaffected.

The tests are much clearer and faster though :)